p2p/net/conn/dial_test.go

```diff
@@ -397,3 +399,221 @@ func TestFailedAccept(t *testing.T) {
 	c.Close()
 	<-done
 }
+
+func TestHangingAccept(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	p1 := tu.RandPeerNetParamsOrFatal(t)
+
+	l1, err := Listen(ctx, p1.Addr, p1.ID, p1.PrivKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p1.Addr = l1.Multiaddr() // Addr has been determined by kernel.
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		con, err := net.Dial("tcp", l1.Addr().String())
+		if err != nil {
+			t.Error("first dial failed: ", err)
+		}
+		// hang this connection
+		defer con.Close()
+
+		// ensure that the first conn hits first
+		time.Sleep(time.Millisecond * 50)
+
+		con2, err := net.Dial("tcp", l1.Addr().String())
+		if err != nil {
+			t.Error("second dial failed: ", err)
+		}
+		defer con2.Close()
+
+		err = msmux.SelectProtoOrFail(SecioTag, con2)
+		if err != nil {
+			t.Error("msmux select failed: ", err)
+		}
+
+		_, err = con2.Write([]byte("test"))
+		if err != nil {
+			t.Error("con write failed: ", err)
+		}
+	}()
+
+	c, err := l1.Accept()
+	if err != nil {
+		t.Fatal("connections after a failed accept should still work: ", err)
+	}
+
+	c.Close()
+	<-done
+}
+
+func TestConcurrentAccept(t *testing.T) {
```

Comment:
pls write a doc explaining a bit of what this test does. not super intuitive. something like:

```
This test kicks off N (=300) concurrent dials, which wait d (=20ms) seconds before failing.
That wait holds up the handshake (multistream AND crypto), which will happen BEFORE
l1.Accept() returns a connection. This test checks that the handshakes all happen
concurrently in the listener side, and not sequentially. This ensures that a hanging dial
will not block the listener from accepting other dials concurrently.
```